[47-BUG] Re : 약속 제안에 가능한 시간 응답 체크

cpplovelove opened the pull request:
## 작업 목록
- [x] 약속 제안 가능 일자 10개 이상 선택 못하도록 처리
- [x] 시간 생성 시에 24시 기준 -> min/maxTime으로 지정해준 시간블록 기준으로 timeTable 저장합니다.
- [x] 약속 제안 id로 조회 시, 약속 제안 timeTable 조회 시에 약속 시간 응답 가능한 날짜와 함께 리턴하도록 수정
- [x] 약속 제안 userId 로 조회 시에 응답한 인원 수와 함께 리턴하도록 수정
- [x] promises/user 확정된 약속 조회 시에 응답 멤버 리스트 추가


## 세부 내용
- 🔗 https://www.notion.so/jalynne/Promising-f4a4b20b38ff4188a3b5d6ce28904563?v=6c89e4aeb22e444b8bda954a65c5a91e 에 수정된 사항 문서로 반영하고 있습니다. 추가 피드백 댓글로 남겨주시면 더 수정 반영하도록 하겠습니다!


## 논의 사항
- min/maxTime 범위 시간 블록 넘어오면 minTime부터 시작으로 카운트해서 타임 생성해주는 형식으로 구현했습니다. front에서 넘겨주는 블록은 maxTime-minTime구간의 시간 블록이라고 가정했어서 minTime 미만 오류는 나지 않는다고 생각했습니다. maxTime 넘길 경우에 badRequest Exception 넘겨주는데 이걸 에러로 봐도 괜찮을 지 ? 살짝 고민됩니다. . . 


## 연결된 이슈
- #47

cpplovelove committed:
refactor: 약속 제안생성 시에 10일 이상 응답시 에러처리

- 응답되는 availDate 저장하고 약속 제안 생성 시에 10일 이상 응답하는
  경우 에러 처리합니다

cpplovelove committed:
refactor: 약속 제안 조회 시에 가능 날짜 같이 리턴

- avilaDate 배열 같이 리턴해서 제안 약속에 대한 정보 알 수 있도록 함

cpplovelove committed:
refactor: 약속 제안 리스트 조회 시에 선택 가능 날짜 조회

- 약속 제안 리스트 조회시에 availDate 같이 리턴

cpplovelove committed:
refactor: 약속 제안 조회 시 응답 인원 수 조회

- memberCount로 약속 제안에 응답한 사용자 수를 함께 리턴합니다.

cpplovelove committed:
refactor: 약속 제안 min/maxTime 체크

- min/max에 따른 블록이 넘어왔을 때에 시간 저장합니다.
- max 타임을 넘었을 경우 BadRequestException 던집니다

binimini commented:
PR 내용 확인해보겠습니당~
minTime 미만은 일단 고려하지 않아야할 것 같구 제 생각에는 maxTime을 넘길 경우에는 두 가지 처리 방식이 가능할 것 같아요
1. controller이나 service 단에서 minTime-maxTime 사이의 최대 개수 (boolean 배열 최대 크기) 를 먼저 계산해서 검증하고 예외처리
2. 좀 더 널널하게 처리해주고 싶다면 maxTime이 넘는 블록부터 무시 (최대 개수까지만 처리)

아무래도 저희 서비스 핵심 로직이다보니 첫번째 방식이 확실할 것 같긴한데 두 번째 방식도 가능할 것 같네요!

cpplovelove commented:
1번 방식 좋은 것 같아요 나머지 피드백도 같이 올려주시면 수정해놓겠습니다!

binimini commented on controllers/promising-controller.ts:
```diff
@@ -19,43 +19,56 @@ import { EventTimeResponse } from '../dtos/event/response';
-    return res.status(200).send(promisingResponse);
+  async getPromisingById(@Param('promisingsId') promisingId: number, @Res() res: Response) {
+    if (!promisingId) throw new ValidationException('');
+    let promisingResponse: any  = await promisingService.getPromisingInfo(promisingId);
+
+    const promisingDateResponse= await promisingDateService.findDatesById(promisingId);    
```
변경되지 않는 변수같은데 const 사용해주시는게 더 좋을 거 같아요. any 타입도 웬만하면 지양해주세요!
그리고 promisingDate를 따로 받는데 promising과 연관되어있으면 쿼리시에 포함해서 조회할 수 있지 않나요? 
따로 서비스로 분리한 이유가 있나용? 🤔 아 다른 곳에서 promisingDate 추가해주는 처리 공통적으로 해주기 위해서인가요?

cpplovelove replied:
커밋하기 전에 한번 더 확인못해서 죄송해요..! 원래 availDate만 리턴해주는 함수 만드려다 promising리스트 조회 시에 같이 리턴해주는 걸로 돼서 기존에 작성해둔 함수 재사용하려다 그렇게 된 걸로 기억합니다..! 데이트 조회 함수 그대로 두고 저 부분은 한꺼번에 조회해올 수 있도록 수정하겠습니다!

binimini commented on services/promise-date-service.ts:
```diff
@@ -0,0 +1,31 @@
+
+        const promisingDateList = promisingDateResponse.map(promisingDateResponse => promisingDateResponse.date)
+        console.log(promisingDateList)
+        return promisingDateList
```
콘솔 코드 삭제해주세요~ 🙃

binimini commented on services/promising-service.ts:
```diff
@@ -64,7 +64,11 @@ class PromisingService {
+      
       if (Object.values(ownPromisingIdList).indexOf(promisingInfo.id) > -1)
         promisingInfo.isOwn = true;
       else promisingInfo.isOwn = false;
```
이 부분 처리 어떤 내용인지 설명 부탁드려도 될까요?
eventId를 undefined 처리 하는 이유도 같이 말씀해주시면 좋을거같아요!

cpplovelove replied:
undefined 예전에 쿼리 조회 방법 잘못찾고 작성해 둔 부분인데 수정해두겠습니다!
윗 부분은 promising에서 약속 제안에 응답한 인원 수 조회하는 부분입니다 !

binimini commented on utils/time.ts:
```diff
@@ -56,22 +59,27 @@ const timeUtil = {
     const dayTime = new Date(day);
-
-    const time = unit * 60;
+    const {minTime,maxTime} = promising;
```
promising 자체를 넘겨주게 되면 유틸 함수의 활용 폭이 좁아질 수 있을거같아서 
필요한 minTime, maxTime만 넘겨주는 식이 더 좋을 것 같아요!

cpplovelove replied:
음 좋아요 ! 요것두 수정하겠습니다!

binimini commented on services/time-service.ts:
```diff
@@ -4,12 +4,13 @@ import timeUtil from '../utils/time';
 
 class TimeService {
-    async create(eventInfo: EventModel, timeInfo: TimeRequest) {
+    async create( promising:PromisingModel, eventInfo: EventModel, timeInfo: TimeRequest) {
         const responseList: Array<TimeModel> = []
-        const resultList: Array<TimeResponse> = timeUtil.boolean2Time(timeInfo) as Array<TimeResponse>
+        const resultList: Array<TimeResponse> = timeUtil.boolean2Time(timeInfo,promising) as Array<TimeResponse>
 
         if (resultList.length == 0) {
             eventService.updateIsAbsent(eventInfo.id, true);
```
사용자가 응답하는 date가 promising의 availDate 내에 있는지 확인하는 로직 구현되어 있나요?
제가 못찾는 건지 관련 로직이 보이지 않네요 😥

cpplovelove replied:
시간만 체크하고 timeTable에 date 체크를 빼먹었네요...지적해주셔서 감사합니다┏ (゜ω゜)=☞ 추가해두겠습니다!

binimini reviewed (commented):
리뷰 달아놨습니당~
쿼리 한 번 내에 포함될 수 있는 로직이 여러 서비스에 분할되긴 했는데
쿼리 자체가 늘어나는 반면에 다시 활용하긴 좋을 것 같아 고민이 되네요 😮
일단은 재사용이 가능한 방향으로 구현해도 좋을 것 같아요!

cpplovelove committed:
refactor: 불필요한 코드 삭제

- 불필요한 console 및 sequelize 코드 삭제

cpplovelove committed:
refactor: 함수 인자 수정

- 시간 단위 리스트를 날짜 데이터로 변환하는 getDateList 함수 인자값 수정

cpplovelove committed:
refactor: 제안 중인 약속조회 로직 수정

- promisingDate 연관으로 가져오도록 변경

cpplovelove committed:
fix: 약속 제안 시간 응답 체크 로직 수정

- timeUtil에서 checkTImeResponseList함수를 통해서 응답 날짜가 응답 날짜 리스트에 들어가 있는지 체크
- 기존에 maxTime 초과 로직 삭제 -> checkTimeResponseList 에서 같이 검사

binimini committed:
refactor: 약속 제안 생성 관련 수정

- 요청시 promising 생성 정보 필터링 로직 추가
- Event 관련 데이터 응답에서 삭제
- 응답시 Date string 변환 처리 추가
- promise-date-service.ts -> promising-date-service.ts
- time util PromisingDate Service 의존성 제거

binimini commented:
몇몇 보이는 사항들을 수정했습니당~ 😀 뭔가 많이 바뀐 것 같지만 사실 prettier 적용이에요ㅎㅎ
- 요청시에 ownerId와 같은 모델에서의 필드가 직접적으로 주입되게 되면 DB에 영향을 줄 수 있어서 
  필터링하는 controller단 처리가 없어 로직을 추가했습니다.
(`{..promisingInfo} = req`와 같이 처리하면 명시하지 않고 추가적으로 들어온 필드도 같이 넘겨지게 됩니당)
- 약속 제안 생성시 Event 관련 데이터를 삭제했습니다.
- 응답에서 Date 자체에서 string 변환 처리를 추가했습니다.
- timeUtil에서 서비스를 의존하지 않도록 수정했습니다.
- `promise-date-service.ts` -> `promising-date-service.ts`로 파일 네이밍을 수정했습니다. 

시간 변환이 약간 안맞는 것 같기도 해서 저 약속 제안 타임 테이블 반환형식 수정하는 김에 한 번 더 확인해보겠습니다! 
머지할게용 수고하셨습니다!🙃

binimini committed:
Merge branch 'dev' of https://github.com/YAPP-Github/20th-Android-Team-1-BE into bug-47-re-time-response
